### PR TITLE
Performant Reshard

### PR DESCRIPTION
**Pull request by ntarafdar**
This performs a Reshard operation where it goes from a tensor sharded in one particular way and outputs a tensor sharded in another way.

Currently we emulate this by doing sharded_tensor -> sharded_to_interleaved -> interleaved_to_sharded

This replaces the above by doing this in one operation. This should reduce the dispatch overhead by doing it in one op.
Furthermore this will reduce a memcpy as the original sharded_tensor is directly being written from input L1 to output L1 of the newly sharded tensor (instead of making an intermediate interleaved tensor in between).

The one limitation of this is for tensors in Row Major layout the last dimension of the shard_spec for the input and output tensors need to be the same as currently that is being used to get the stick size which needs to be the same.
This can be addressed in a future PR.

This has been added at ttnn and tt_lib.

@arakhmati @eyonland please check the ttnn implementation and adding of the ttnn tests
@tt-aho please look into the sharded_op_multi_core.cpp file which is the host, key thing there look at how I give the kernel which is reshard_reader.cpp the sharding information (cores and address ranges). I used something similar in the fast dispatch kernel to actually read sharded buffers.

@mywoodstock @tt-nshanker please look at the tests I added, we can sync on some more conv related tests we can look at. I have added the tests at both ttnn and tt_lib levels.

Note: This also adds a fix for overlapping shards. This was not previously supported and adds some host-side compute. This has added some host-side time for a resnet perf test.

Passes most tests here:
https://github.com/tenstorrent-metal/tt-metal/actions/runs/8014445456

Fast dispatch tests failed because it lost connection with runner, so only re-running fast dispatch tests: 
https://github.com/tenstorrent-metal/tt-metal/actions/runs/8019920776

CC: @davorchap

### CODE REVIEW

**Review by arakhmati — APPROVED** [review]

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py [hunk 1]
```diff
@@ -1481,3 +1482,1604 @@ def test_sharded_concat_heads(
     passing, output = comp_pcc(pt_out, tt_out)
     logger.info(output)
     assert passing
+
+
+@pytest.mark.parametrize(
```
Comment (tt-aho) [inline tests/tt_eager/python_api_testing/unit_testing/test_sharded.py:1487]: Are all these tests just duplicated? Or what's the difference?

File: tt_eager/tt_dnn/op_library/sharded/multi_core/sharded_op_multi_core.cpp [hunk 2]
```diff
@@ -502,6 +503,218 @@ operation::ProgramWithCallbacks sharded_to_interleaved_multi_core(
...
+    else {
+        small_to_big_ratio = input_buffer->num_pages()/output_shard_to_host_mapping.size();
+        num_pages = output_shard_to_host_mapping.size();
+        input_smaller = false;
+    }
+
+// First get output_core to vector< pair<input_core, input_page> (num_pages_in_output)
+    std::unordered_map <CoreCoord, std::vector<std::pair<CoreCoord, uint32_t> > > output_core_to_vector_input_core_page;
+
+//    for(uint32_t outer_page_id = 0; outer_page_id < num_pages; outer_page_id++) {
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/sharded/multi_core/sharded_op_multi_core.cpp:536]: Remove dead code

File: tt_eager/tt_dnn/op_library/sharded/kernels/dataflow/reshard_reader.cpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+//        DPRINT << "range_id: " << range_id << ENDL();
+//        DPRINT << "core_id_x: " << core_id_x <<  ENDL();
+//        DPRINT << "core_id_y: " << core_id_y <<  ENDL();
+//        DPRINT << "offset: " << offset << ENDL();
+//        DPRINT << "size: " << size << ENDL();
+        noc_async_read(noc_address, l1_write_addr, size);
+        l1_write_addr+=size;
+
+    }
+    cb_push_back(shard_cb, num_output_pages);
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/sharded/kernels/dataflow/reshard_reader.cpp:37]: You're missing a read barrier

File: tt_eager/tt_dnn/op_library/sharded/kernels/dataflow/reshard_reader.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    for(uint32_t range_id = 0; range_id <num_ranges; range_id++) {
+        uint32_t core_id_x = get_arg_val<uint32_t>(arg_index++);
+        uint32_t core_id_y = get_arg_val<uint32_t>(arg_index++);
+        uint32_t offset = get_arg_val<uint32_t>(arg_index++);
+        uint32_t size = get_arg_val<uint32_t>(arg_index++);
+        uint64_t noc_address = get_noc_addr(core_id_x, core_id_y,
+                                        input_shard_addr + offset);
+//        DPRINT << "range_id: " << range_id << ENDL();
+//        DPRINT << "core_id_x: " << core_id_x <<  ENDL();
+//        DPRINT << "core_id_y: " << core_id_y <<  ENDL();
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/sharded/kernels/dataflow/reshard_reader.cpp:30]: Remove dead code

File: tests/ttnn/unit_tests/operations/test_core.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Comment (tt-aho) [inline tests/ttnn/unit_tests/operations/test_core.py:1]: Maybe name this file something better? test_core doesn't seem very descriptive to what this file is testing?

File: tt_eager/tt_dnn/op_library/sharded/sharded_op.hpp [hunk 6]
```diff
@@ -141,6 +181,21 @@ inline Tensor sharded_to_interleaved(
         .at(0);
 }
 
+inline Tensor reshard(const Tensor &input_tensor, const MemoryConfig &output_mem_config) {
+    TT_FATAL(input_tensor.shard_spec().has_value());
+    TT_FATAL(output_mem_config.is_sharded());
+
+    return operation::run(
+               Reshard{
+                   .input_mem_config = input_tensor.memory_config(),
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/sharded/sharded_op.hpp:190]: Any reason why input_mem_config is an attribute if input tensor is an input that we take memconfig from?

Pr is this for combining with the other struct?

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py [hunk 7]
```diff
@@ -1481,3 +1482,1604 @@ def test_sharded_concat_heads(
...
+    pt_out = torch.transpose(in0, -3, -2).reshape([B, 1, seq_len, num_heads * head_dim])
+
+    tt_out = tt2torch_tensor(output_t)
+
+    passing, output = comp_pcc(pt_out, tt_out)
+    logger.info(output)
+    assert passing
+
+
+@pytest.mark.parametrize("debug", [False])
```
Comment (tt-aho) [inline tests/tt_eager/python_api_testing/unit_testing/test_sharded.py:2953]: Remove?

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py [hunk 8]
```diff
@@ -1481,3 +1482,1604 @@ def test_sharded_concat_heads(
...
+@pytest.mark.parametrize(
+    "input_shape, input_layout, input_shard_grid,  input_shard_shape, input_shard_orientation, input_sharding_scheme, output_shard_grid, output_shard_shape, output_shard_orientation, output_sharding_scheme ",
```
Comment (tt-aho) [inline tests/tt_eager/python_api_testing/unit_testing/test_sharded.py:2955]: Are there tests for width / height sharding with reshard? Currently all the ones I saw were only block sharding. Are there also program_cache tests for reshard?

**Review by tt-aho — CHANGES_REQUESTED** [review]

**Comment by tt-nshanker** [comment]
@tarafdarTT Thanks for this change. The TTNN conv unit test uses functions defined in `tt_eager/tt_dnn/op_library/sliding_window_op_infra/tt_py_composite_conv.py` to copy data from interleaved to sharded and vice versa. Can you please update the functions, `conv_input_interleaved_to_sharded` and `conv_output_sharded_to_interleaved` and run the CI? Thanks

File: tt_metal/impl/buffers/buffer.hpp [hunk 9]
```diff
@@ -197,6 +197,9 @@ class Buffer {
     }
 
     CoreCoord get_core_from_dev_page_id(uint32_t dev_page_id) const {
+        if(dev_page_id >= dev_page_to_core_mapping_.size()) {
+            std::cout << "WTF " << std::endl;
```
Comment (xanderchin) [inline tt_metal/impl/buffers/buffer.hpp:201]: 😅 I'm guessing you probably didn't mean to leave this in

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py (hunk 1 above)
Comment (tarafdarTT) [reply tests/tt_eager/python_api_testing/unit_testing/test_sharded.py:1487]: duplicated with the ttnn tests?

the ttnn tests and the tt_lib tests are the same for the most part. I did add a few more height and width sharded tests at the ttnn level.

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py (hunk 8 above)
Comment (tarafdarTT) [reply tests/tt_eager/python_api_testing/unit_testing/test_sharded.py:2955]: oh I just added some height and width sharded tests at the ttnn level

File: tests/ttnn/unit_tests/operations/test_core.py (hunk 5 above)
Comment (tarafdarTT) [reply tests/ttnn/unit_tests/operations/test_core.py:1]: the test name for now is consistent with the ttnn file the op is in. 
ttnn/operations/core.py

I was thinking that to this file we could add the other core operations

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py (hunk 1 above)
Comment (tt-aho) [reply tests/tt_eager/python_api_testing/unit_testing/test_sharded.py:1487]: I meant like it seems like the tests that were originally in this file were copy pasted. Ex `test_sharded_tile` now shows up twice in this file

Comment (tarafdarTT) [reply tests/tt_eager/python_api_testing/unit_testing/test_sharded.py:1487]: oooh hehe whoops , ya that's a rebasing kerfuffle. will fix this.

File: tt_eager/tt_dnn/op_library/sharded/sharded_op.hpp (hunk 6 above)
Comment (tarafdarTT) [reply tt_eager/tt_dnn/op_library/sharded/sharded_op.hpp:190]: Ya this was meant to combine later but I think I will remove it for now, and when combining I can add it back.

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py (hunk 8 above)
Comment (tarafdarTT) [reply tests/tt_eager/python_api_testing/unit_testing/test_sharded.py]: I will add some program_cache tests

File: tt_metal/impl/buffers/buffer.hpp (hunk 9 above)
Comment (tarafdarTT) [reply tt_metal/impl/buffers/buffer.hpp]: lolol whoops, it has been removed :)

File: tt_eager/tt_dnn/op_library/sharded/kernels/dataflow/reshard_reader.cpp (hunk 3 above)
Comment (tt-aho) [reply tt_eager/tt_dnn/op_library/sharded/kernels/dataflow/reshard_reader.cpp:29]: read barrier should come before push_back for correctness (even though consumer is no-op)

**Review by tt-aho — COMMENTED** [review]
One question I have is why the inference times for resnet needed to be bumped up? This seems like a significant degradation going form 0.015->0.025?

**Comment by mywoodstock** [comment]
> One question I have is why the inference times for resnet needed to be bumped up? This seems like a significant degradation going form 0.015->0.025?

0.015 -> 0.025 is a huge regression. Where are these numbers from?

**Comment by tarafdarTT** [comment]
> One question I have is why the inference times for resnet needed to be bumped up? This seems like a significant degradation going form 0.015->0.025?

The only thing for resnet that this touches are the changes of the shard mappings to support uneven mappings as well as clean up Block sharding, which may also fix other bugs observed for block sharding.
This is now correct for at least for the shapes that I tested, I can look at the host side also though, off the top of my head I see a vector push back (I use that because for uneven pages per core we don't know exactly, but I do know maximum and can allocate that).

**Comment by tarafdarTT** [comment]
> > One question I have is why the inference times for resnet needed to be bumped up? This seems like a significant degradation going form 0.015->0.025?
> 
> 0.015 -> 0.025 is a huge regression. Where are these numbers from?

See below.

**Review by boris-drazic — APPROVED** [review]

**Review by tt-aho — APPROVED** [review]
Approving since I will be away and not sure how to just remove my request so I don't block.

@mywoodstock you can request changes if you see any blockers

**Comment by tarafdarTT** [comment]
> Approving since I will be away and not sure how to just remove my request so I don't block.
> 
> @mywoodstock you can request changes if you see any blockers

thanks @tt-aho , the perf issues should be fixed.

**Comment by mywoodstock** [comment]
> > Approving since I will be away and not sure how to just remove my request so I don't block.
> > @mywoodstock you can request changes if you see any blockers
> 
> thanks @tt-aho , the perf issues should be fixed.

Yeah, my main concern was about using `get_arg_val` in the main loop -- perf wise.

**Comment by tarafdarTT** [comment]
> > > Approving since I will be away and not sure how to just remove my request so I don't block.
> > > @mywoodstock you can request changes if you see any blockers
> > 
> > 
> > thanks @tt-aho , the perf issues should be fixed.
> 
> Yeah, my main concern was about using `get_arg_val` in the main loop -- perf wise.

sure, to clarify the perf issues we saw were not related to that, as that was not using reshard. it is in the host code.
could you point me to an example that uses the mem address for variadic arguments.

**Comment by mywoodstock** [comment]
> > > > Approving since I will be away and not sure how to just remove my request so I don't block.
> > > > @mywoodstock you can request changes if you see any blockers
> > > 
> > > 
> > > thanks @tt-aho , the perf issues should be fixed.
> > 
> > 
> > Yeah, my main concern was about using `get_arg_val` in the main loop -- perf wise.
> 
> sure, to clarify the perf issues we saw were not related to that, as that was not using reshard. it is in the host code related to the formation of the buffer. could you point me to an example that uses the mem address for variadic arguments.

Yes, I was concerned in a unrelated way to the e2e drop :)
I was suggesting creating a 1d tensor (config tensor), with all the args you need, and just access them like a regular tensor.

**Comment by tarafdarTT** [comment]
> > > > > Approving since I will be away and not sure how to just remove my request so I don't block.
> > > > > @mywoodstock you can request changes if you see any blockers
> > > > 
> > > > 
> > > > thanks @tt-aho , the perf issues should be fixed.
> > > 
> > > 
> > > Yeah, my main concern was about using `get_arg_val` in the main loop -- perf wise.
> > 
> > 
> > sure, to clarify the perf issues we saw were not related to that, as that was not using reshard. it is in the host code related to the formation of the buffer. could you point me to an example that uses the mem address for variadic arguments.
> 
> Yes, I was concerned in a unrelated way to the e2e drop :) I was suggesting creating a 1d tensor (config tensor), with all the args you need, and just access them like a regular tensor within the kernel.

is there a way for this to be made at the host?

**Comment by mywoodstock** [comment]
> > > > > > Approving since I will be away and not sure how to just remove my request so I don't block.
> > > > > > @mywoodstock you can request changes if you see any blockers
> > > > > 
> > > > > 
> > > > > thanks @tt-aho , the perf issues should be fixed.
> > > > 
> > > > 
> > > > Yeah, my main concern was about using `get_arg_val` in the main loop -- perf wise.
> > > 
> > > 
> > > sure, to clarify the perf issues we saw were not related to that, as that was not using reshard. it is in the host code related to the formation of the buffer. could you point me to an example that uses the mem address for variadic arguments.
> > 
> > 
> > Yes, I was concerned in a unrelated way to the e2e drop :) I was suggesting creating a 1d tensor (config tensor), with all the args you need, and just access them like a regular tensor within the kernel.
> 
> is there a way for this to be made at the host?
> 
> sorry I don't think I've seen too many examples of tensors made at the host as a config for the kernel running on device.
> 
> Edit: Ahh sorry I think I get it now.

One way could be creating the tensor before the op call, but then you will need to add another input tensor. Other way you can create a tensor with owned buffer on host in the op.

**Review by mywoodstock — APPROVED** [review]

**Comment by tt-aho** [comment]
@mywoodstock is it really a perf issue to call get_arg_val in a loop? This is just a local l1 lookup with addition/bit shift, and each arg is only read once.

Wouldn't this the same perf-wise as passing in an input tensor, and then reading from that tensor's L1 since you'd just be offsetting the tensor's base address as well?

Or if the offset calculation of get_arg_val is inefficient, we could do the arg address calculation ourselves to minimize additions/bit shifts since we know the offset sizes are equal.

**Comment by mywoodstock** [comment]
> @mywoodstock is it really a perf issue to call get_arg_val in a loop? This is just a local l1 lookup with addition/bit shift, and each arg is only read once.
> 
> Wouldn't this the same perf-wise as passing in an input tensor, and then reading from that tensor's L1 since you'd just be offsetting the tensor's base address as well?
> 
> Or if the offset calculation of get_arg_val is inefficient, we could do the arg address calculation ourselves to minimize additions/bit shifts since we know the offset sizes are equal.

AFAIR its the overhead of these func calls (offset calc etc) which would be the concern. Not sure how much though.

**Comment by tarafdarTT** [comment]
> > @mywoodstock is it really a perf issue to call get_arg_val in a loop? This is just a local l1 lookup with addition/bit shift, and each arg is only read once.
> > Wouldn't this the same perf-wise as passing in an input tensor, and then reading from that tensor's L1 since you'd just be offsetting the tensor's base address as well?
> > Or if the offset calculation of get_arg_val is inefficient, we could do the arg address calculation ourselves to minimize additions/bit shifts since we know the offset sizes are equal.
> 
> AFAIR its the overhead of these func calls (offset calc etc) which would be the concern. Not sure how much though.

Hmm, if it's fuzzy, how about we address this in a separate PR. this should still be an improvement over sharded_to_interleaved and interleaved_to_sharded as there is no host intervention in between.

**Comment by tt-aho** [comment]
> > @mywoodstock is it really a perf issue to call get_arg_val in a loop? This is just a local l1 lookup with addition/bit shift, and each arg is only read once.
> > Wouldn't this the same perf-wise as passing in an input tensor, and then reading from that tensor's L1 since you'd just be offsetting the tensor's base address as well?
> > Or if the offset calculation of get_arg_val is inefficient, we could do the arg address calculation ourselves to minimize additions/bit shifts since we know the offset sizes are equal.
> 
> AFAIR its the overhead of these func calls (offset calc etc) which would be the concern. Not sure how much though.

I see. This is beyond the scope of this pr but I think it's possible to make this lookup much easier without these offsets.

Assuming all your arg data types are the same, you could have a `get_arg_ptr<data_type>` func that just does `return (volatile tt_l1_ptr T*)(L1_ARG_BASE)`;
so that way arg lookups are just pointer indexing and there are no offset calculations.
This is useful if args are only used once, but if you need to reuse args we could read those ones into non-volatile variables.

**Review by kmabeeTT — APPROVED** [review]

**Pull request merged**